### ISSUE TITLE
SQL api for system tables

### ISSUE THREAD

**Issue by alex-dukhno** [alex-dukhno/database#114]
Right now, despite the fact that table information is written to `system` namespace (effectively schema) user can't retrieve data out of it.

SQL to reproduce:
```SQL
create schema schema_name;
create table schema_name.table_name (col_1 smallint, col_2 integer, col_3 bigint);
select * from system.columns;
```

Expected result:
```
sql_type | name
---------+-----
smallint | col_1
smallint | col_2
smallint | col_3
(3 rows)
```

Actual result:
```
select * from system.columns;
--
(1 row)
```

**Comment by vkulichenko** [comment]
Alex, I would like to pick this one up. Can you please assign it to me?

**Comment by vkulichenko** [comment]
@alex-dukhno Pull request created. Please let me know if you have any questions about the implementation.

**Comment by vkulichenko** [comment]
PR recreated with the "DROP TABLE" support and additional functional tests.

**Comment by vkulichenko** [comment]
@alex-dukhno Hi Alex. It looks like I will have to revisit the implementation due to the latest changes in the storage. So I will close the current PR and open a new one when I'm ready.

I also wanted to hear your thoughts on how this should be implemented. The main problem here is that `system.columns` is a regular table which can be queried via SQL, but at the same time, it stores internal information like the type of a column (`SqlType`). What is the best format for it?

One option is to convert `SqlType` to a string, but that doesn't work well for the `table_columns` method, which returns `ColumnDefinition`s.

My suggestion is to add one more `Datum` type, which can store anything that implements `Serialize`, `Deserialize`, and `Display` traits. Such values can be stored in binary form, and still be showed as strings when queried by SQL. Internally, we can just use deserialized values.

This type doesn't have to be exposed to the public as a SQL type but will be useful for such internal purposes. I believe this will be relevant for any meta table like this, should they be added in the future.

WDYT? If this approach makes sense to you, I can work on the implementation.

**Comment by alex-dukhno** [comment]
Hi Valentin, 

I'll use this comment more to dump my ideas, rather than make some decision (I believe I need some more info to gather before I will be satisfied with state of `storage`)

1. Currently, latest on disk representation is influenced by prototype of `QueryProcessor` from https://github.com/alex-dukhno/database/pull/183, where Materialize were used as a reference. I am not very comfortable to go with current state of it. I've created a research/discussion issue https://github.com/alex-dukhno/database/issues/221
The good news is that we have some movement in this direction because I don't like binary departed by special byte `|`
1. The task on my list for the next couple of days is to work on `FrontendStorage` API. Previously, schema, table, column existence was checked inside `storage`, now it was moved into `sql_engine`. It heavily influence error handling inside storage. There is a possibility that it also influence the next my point
1. After your question on design and suggestion how to store `SqlType` I started thinking on adding `system.sql_types` table with `id` (`smallint` / `i16`) and `name` (`varchar`) columns. And instead of using `Serialize`/`Deserialize` trait just convert `SqlType` into `i16`

Of cause we can discuss here or in [discord channel](https://discord.gg/PUcTcfU) the third option

Thanks

**Comment by vkulichenko** [comment]
Alex, should I postpone this ticket until these major changes are completed?

Your suggestion generally makes sense. An issue that I see here is that `select * from system.columns` will return integers rather than type names. A user will need to do a join to get a meaningful result, which might seem like an overkill for such a use case.

What I was trying to suggest above is a more generic mechanism for storing arbitrary types for this kind of internal purposes. I do realize it has downsides as well though.

Bottom line - I think we need to decide on the API first, and then jump into implementation.

**Comment by alex-dukhno** [comment]
> should I postpone this ticket until these major changes are completed?

I think yes, sorry for that.

> Bottom line - I think we need to decide on the API first, and then jump into implementation.

I need time to think through. I don't have an idea that sounds like solution, right now I have only work arounds.

**Comment by alex-dukhno** [comment]
ANSI SQL has `information schema` and PG docs https://www.postgresql.org/docs/9.5/information-schema.html
